libvncclient/rfbproto.c

```diff
@@ -680,6 +680,60 @@ rfbPowM64(uint64_t b, uint64_t e, uint64_t m)
   return r;
 }
 
+static rfbBool
+HandleUltraMSLogonIIAuth(rfbClient *client)
+{
+  uint8_t gen[8], mod[8], resp[8], pub[8], priv[8];
+  uint8_t username[256], password[64], key[8];
+  rfbCredential *cred;
+
+  if (!ReadFromRFBServer(client, (char *)gen, sizeof(gen))) return FALSE;
+  if (!ReadFromRFBServer(client, (char *)mod, sizeof(mod))) return FALSE;
+  if (!ReadFromRFBServer(client, (char *)resp, sizeof(resp))) return FALSE;
+
+  if(!dh_generate_keypair(priv, pub, gen, sizeof(gen), mod, sizeof(priv))) {
+      rfbClientErr("HandleUltraMSLogonIIAuth: generating keypair failed\n");
+      return FALSE;
+  }
+
+  if(!dh_compute_shared_key(key, priv, resp, mod, sizeof(key))) {
+      rfbClientErr("HandleUltraMSLogonIIAuth: creating shared key failed\n");
+      return FALSE;
+  }
+
+  if (!client->GetCredential)
+  {
+    rfbClientLog("GetCredential callback is not set.\n");
+    return FALSE;
+  }
+  rfbClientLog("WARNING! MSLogon security type has very low password encryption! "\
```

Comment:
This is admittedly copy/pasted, I haven't put too much thought into whether it can still be considered the case.

Reply:
MSlogon should always be used together with the dsm plugins so the user/password credentials are verified insite the AES stream. Without the simple encryption isn't secure enough for usage over the net.